specs/capri/Monitoring Design.md

```diff
@@ -0,0 +1,742 @@
+
+*CONTROLLER*
+
+Controller handles POST and GET requests from API server.
+
+1.  The controller will translate requests from the GET API to a PromQL call,
+    get the data from Prometheus and send back to the API server.
+
+2.  The controller gets POST request from API server, sends requests to Dock to
+    generate metrics from the metrics drivers. Metrics collected this way will
+    be converted and pushed to the exporters that are configured.
+
+*PROMETHEUS*
+
+1.  This is installed and configured by default, can be optionally removed.
+
+2.  This can collect data in two ways and stored it to the TSDB
+
+    1.  The exporter in the dock will send data to the Prometheus. Here, we have
+        2 options, have prototyped both, will select one of them
+
+        1.  Use the Prometheus API to send metrics to the push gateway (after a
+            collection happened in response to a POST call)
+
+        2.  Configure the text collector of the Node exporter in Prometheus (the
+            collector writes the text file with the metrics and the text
+            collector pushes them to Prometheus)
+            <https://github.com/prometheus/node_exporter#textfile-collector>
+
+    2.  Prometheus is configured to scrape the exporter end points periodically
+        (PULL mode)
+
+3.  OpenSDS will ship with guidelines and example configuration of Prometheus
+    for both modes
+
+*DOCK*
+
+1.  A collector module is added to the dock, which collects metrics in response
+    to the POST API call
+
+2.  Collected metrics are then sent to all the exporters that are configured
+
+3.  The exporters do 2 things
+
+    1.  Conversion
+
+    2.  Sending to the requisite endpoint (Prometheus push gateway, KAFKA etc.)
+
+4.  The exporters are designed to be extendable...to be able to convert and send
+    data to a custom 3rd party endpoint
+
+5.  The Prometheus exporter also exposes a HTTP endpoint so it can be scraped by
+    the Prometheus server (in PULL mode)
+
+*DRIVER*
+
+1.  The metrics driver for a particular storage can be added to an existing
+    volume driver, or it can be an independent Metrics driver. The metrics driver interface is defined below
+    ```go
+        // MetricDriver is an interface for exposing metrics collection operation of different components (Volume/CPU/Memory etc)
+        type MetricDriver interface {
+                //Any initialization the Metric driver does while starting.
+                Setup() error
+                //Any operation the Metric driver does while stopping.
+                Unset() error
+                //Every POST operation will trigger a CollectMetrics operation for specific MetricDriver
+                //metricList:- Controller  passed list of metrics .
+                //instanceID:- is the component name/id (VolumeID,CPU_ID)
+                //*model.Metric:-  The array of metrics collected .
+                CollectMetrics([]metricList string,instanceID string) ([]*model.Metric, error)
```

Comment:
s/[]metricList string/metricList []string

Reply:
can you please elaborate, thanks.

Reply:
Because in golang world, array should be expressed as `[]string`. So if I get you right, you should change that type.